cirq-google/cirq_google/study/finite_random_variable.py

```
@@ -0,0 +1,119 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from typing import Any, Iterator
+
+import cirq
+from cirq.study.sweeps import SingleSweep
+
+
+class FiniteRandomVariable(SingleSweep):
+    """A sweep over randomly-sampled values from a finite distribution.
+
+    This can generate a stream of random samples pulled from a finite
+    distribution.  For instance, a coin flip, die roll, or from {1, 0, -1}.
+
+    Note: this is an experimental sweep and is not guaranteed to be
+    future-compatible.
+
+    It is highly recommended that you specify a seed, so that the samples
+    are reproducible.
+    """
+
+    def __init__(
+        self,
+        key: cirq.TParamKey,
+        distribution: dict[float, float],
+        seed: int,
+        length: int,
+        metadata: Any | None = None,
+    ) -> None:
+        """Creates a random sweep for a given key.
+
+        Args:
+            key: sympy.Symbol or equivalent to sweep across.
+            distribution: A dictionary mapping values to their probabilities.
+                Probabilities do not need to sum to 1; they will be
+                normalized by `random.choices`.
+            seed: A seed for the random number generator to ensure
+                reproducibility.
+            length: The number of points in the sweep.
+            metadata: Optional metadata to attach to the sweep.
+        """
+        if not isinstance(distribution, dict) or not distribution:
+            raise ValueError("distribution must be a non-empty dictionary.")
+        if not all(isinstance(k, (int, float)) for k in distribution.keys()):
+            raise ValueError("distribution keys must be numbers.")
+        if not all(isinstance(v, (int, float)) and v >= 0 for v in distribution.values()):
+            raise ValueError("distribution values (weights) must be non-negative numbers.")
```

Should we also raise an error if sum(distribution.values()) is not close to 1.0?

---

I don't think so.  They are really just weights, so if it doesn't sum to one, it all gets normalized by `random.choices` anyway.   This way, it's very easy to specify 1/3 for each one: just {0: 1, 1:1, 2:1}, for instance.